### PR TITLE
add umf split mode to os_memory_provider

### PR DESCRIPTION
**Pull request by lplewa**
split mode splits each allocation based on user provided partitions,
across multiple numa nodes. If no partition defined then each
allocation is split evenly, across all numa available.

Signed-off-by: Łukasz Plewa <lukasz.plewa@intel.com>

### CODE REVIEW

File: src/provider/provider_os_memory.c [hunk 1]
```diff
@@ -25,19 +25,22 @@
...
 
+#ifdef WIN32
+typedef ssize_t SSIZE_T
+#endif
 #define NODESET_STR_BUF_LEN 1024
 
 #define TLS_MSG_BUF_LEN 1024
 
-typedef struct os_last_native_error_t {
+    typedef struct os_last_native_error_t {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 2]
```diff
@@ -25,19 +25,22 @@
...
     int32_t native_error;
     int errno_value;
     char msg_buff[TLS_MSG_BUF_LEN];
 } os_last_native_error_t;
 
 static __TLS os_last_native_error_t TLS_last_native_error;
 
-// helper values used only in the Native_error_str array
+// helper values used only in the Native_error_str arrayq
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 3]
```diff
@@ -230,6 +238,78 @@ static int getHwlocMembindFlags(umf_numa_mode_t mode, int dedicated_node_bind) {
...
 
+static umf_result_t
+validatePartitions(umf_os_memory_provider_params_t *params) {
+
+    if (params->partitions_len == 0) {
+        return UMF_RESULT_SUCCESS;
+    }
+    for (int i = 0; i < params->partitions_len; i++) {
+        int found = 0;
+        if (params->partitions->weight == 0) {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 4]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
     return 0;
 }
 
-static int get_membind(os_memory_provider_t *provider, size_t size) {
+typedef struct membind_t {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c:608]: We do not have comments above internal struct and functions. I added few in this file where it is needed.

File: src/provider/provider_os_memory.c [hunk 5]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+
+    /// Number of pages to allocate
+    size_t pages;
+
+    /// Pages left to bind in current node
+    size_t leftover_bind;
+} membind_t;
+
+/// Advances the memory binding configuration for the next set of pages
+/// If there we have to bind bytes which belongs to single page, to mutiliple nodes.
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 6]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+        return;
+    }
+
+    // Reset the bitmap for next binding
+    hwloc_bitmap_zero(membind->bitmap);
+
+    // Flag to check if binding crosses partition boundaries
+    int bind_border_page = 0;
+    if (membind->leftover_bind != 0) {
+        // if we have move then page lestover from previous bind
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 7]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+        if (rest >= provider->partitions_weight_sum) {
+            bind++;
+            rest -= provider->partitions_weight_sum;
+        }
+
+        // Update the bitmap to include the current node's target
+        hwloc_bitmap_or(membind->bitmap, membind->bitmap,
+                        provider->partitions[membind->node].target);
+
+        // If there current node has to bind less then page
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 8]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+    hwloc_bitmap_zero(membind->bitmap);
+
+    // Flag to check if binding crosses partition boundaries
+    int bind_border_page = 0;
+    if (membind->leftover_bind != 0) {
+        // if we have move then page lestover from previous bind
+        hwloc_bitmap_or(membind->bitmap, membind->bitmap,
+                        provider->partitions[membind->node].target);
+    } else if (membind->rest != 0) {
+        // if we have less then page leftover to bind from previous bind
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 9]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+            membind->node++;
+            assert(membind->node < provider->partitions_len);
+            bind_border_page = 1;
+        }
+    }
+
+    // Update bind size and remainder based on whether the binding crossed a partition boundary
+    if (bind_border_page) {
+        // this means that next page bellong to multiple nodes.
+        // in this case we has to bind this page separatly, and
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: done

[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c [hunk 10]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+        membind->bind_size = membind->page_size;
+        membind->leftover_bind = bind - 1;
+        membind->rest = rest;
+    } else {
+        membind->bind_size = membind->page_size * bind;
+        membind->rest = rest;
+        membind->leftover_bind = 0;
+    }
+    // if processing this node finished move to next one
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Done

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 11]
```diff
@@ -425,6 +425,165 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+TEST_P(testNumaSplit, checkModeSplit) {
+    auto in = GetParam();
+    size_t page_size = sysconf(_SC_PAGE_SIZE);
+
+    umf_os_memory_provider_params_t os_memory_provider_params =
+        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+
+    std::vector<unsigned> numa_nodes = get_available_numa_nodes();
+
+    if (numa_nodes.size() < std::get<0>(in)) {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp:535]: Each test iteration  requires minimum numa nodes.

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 12]
```diff
@@ -425,6 +425,165 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+        GTEST_SKIP_("Not enought numa nodes");
+    }
+
+    ASSERT_EQ(std::get<3>(in).size(), std::get<1>(in)) << "Wrong test in data";
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp:539]: It was sanity check to ensure that i defined correctly test input

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 13]
```diff
@@ -425,6 +425,165 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+
+    size_t size = page_size * std::get<1>(in);
+    umf_result_t umf_result;
+    umf_result = umfMemoryProviderAlloc(os_memory_provider, size, 0, &ptr);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    ASSERT_NE(ptr, nullptr);
+
+    unsigned long maxnode = 1024;
+    unsigned long *nodemask =
+        (unsigned long *)malloc(maxnode); // Allocate the nodemask buffer
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: chatgpt generated comment xD - i will remove it

File: include/umf/providers/provider_os_memory.h [hunk 14]
```diff
@@ -57,12 +57,23 @@ typedef enum umf_numa_mode_t {
...
+    UMF_NUMA_MODE_SPLIT,
     /// The memory is allocated on the node of the CPU that triggered the
     /// allocation. If this mode is specified, nodemask must be NULL and
     /// maxnode must be 0.
     UMF_NUMA_MODE_LOCAL, // TODO: should this be a hint or strict policy?
 } umf_numa_mode_t;
 
+/// @brief partition specification for UMF_NUMA_MODE_SPLIT
+typedef struct umf_numa_split_partition_t {
+    /// weight of partition.
```
Comment (lukaszstolarczuk) [inline include/umf/providers/provider_os_memory.h:71]: perhaps a little more precise here, pls...?

File: include/umf/providers/provider_os_memory.h [hunk 15]
```diff
@@ -57,12 +57,23 @@ typedef enum umf_numa_mode_t {
     /// fulfilled, memory will be allocated from other nodes.
     UMF_NUMA_MODE_PREFERRED,
 
+    /// Allocation will be split evenly across nodes specified in nodemask.
+    /// An extra argument could be passed in params structure to specify other distribution.
```
Comment (lukaszstolarczuk) [inline include/umf/providers/provider_os_memory.h:61]: do you mean partitions arguments in `umf_os_memory_provider_params_t`? these are not `an extra argument` - you could be more precise, pls

File: include/umf/providers/provider_os_memory.h [hunk 16]
```diff
@@ -57,12 +57,21 @@ typedef enum umf_numa_mode_t {
...
+    /// allocation will be splited evenly across nodes specified in nodemask.
+    /// an extra argument could be passed to specify other distribution.
+    UMF_NUMA_MODE_SPLIT,
     /// The memory is allocated on the node of the CPU that triggered the
     /// allocation. If this mode is specified, nodemask must be NULL and
     /// maxnode must be 0.
     UMF_NUMA_MODE_LOCAL, // TODO: should this be a hint or strict policy?
 } umf_numa_mode_t;
 
+///
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply include/umf/providers/provider_os_memory.h]: thx

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:535]: yeah, that I anticipated 😉 I wasn't sure how (technically) it is achieved, but if you say it should work, then I believe you 😄

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:539]: sooo... it should be removed...?

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 17]
```diff
@@ -425,6 +425,165 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
     umfMemoryProviderFree(os_memory_provider, ptr, size);
 }
 
+using numaSplitOut = std::vector<std::vector<unsigned>>;
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:428]: bump

File: src/provider/provider_os_memory.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory.c:608]: I know we don't have, but it's not a good example 😄 It be nice to look at this struct and know what it is for

File: src/provider/provider_os_memory.c [hunk 18]
```diff
@@ -279,14 +356,36 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
     }
     LOG_INFO("established HWLOC NUMA policy: %u", provider->numa_policy);
 
+    result = validate_numa_mode(in_params->numa_mode, emptyNodeset);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect NUMA mode (%u) or wrong params",
+                in_params->numa_mode);
+        return result;
+    }
```
Comment (PatKamin) [inline src/provider/provider_os_memory.c:364]: This code is duplicated, I don't see why.

File: include/umf/providers/provider_os_memory.h (hunk 15 above)
Comment (lplewa) [reply include/umf/providers/provider_os_memory.h]: done

File: src/provider/provider_os_memory.c (hunk 18 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: Copy-paste error done

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 19]
```diff
@@ -425,6 +425,160 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+
+    os_memory_provider_params.numa_list = numa_nodes.data();
+    os_memory_provider_params.numa_list_len = std::get<0>(in);
+    os_memory_provider_params.numa_mode = UMF_NUMA_MODE_SPLIT;
+
+    os_memory_provider_params.partitions = std::get<2>(in).data();
+    os_memory_provider_params.partitions_len = std::get<2>(in).size();
+    initOsProvider(os_memory_provider_params);
+
+    size_t size = page_size * std::get<1>(in);
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: Done

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 20]
```diff
@@ -425,6 +425,160 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+        out = mergeOutVec({{{0}, {0, 1}},
+                           numaSplitOut(3, {1}),
+                           {{1, 2}, {1, 2}},
+                           numaSplitOut(8, {1}),
+                           {{0, 1, 2}},
+                           numaSplitOut(15, {0})});
+
+        ret.push_back({3, 31, in, out});
+
+        in = {{31, 1}, {17, 1}, {14, 2}, {19, 0}, {8, 2}, {5, 1}, {15, 2}};
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp:472]: Done

File: include/umf/providers/provider_os_memory.h (hunk 14 above)
Comment (lplewa) [reply include/umf/providers/provider_os_memory.h]: Done

**Comment by bratpiorka** [comment]
please add one test for integer overflow where one weight = 1 and second = max int

**Review by PatKamin — APPROVED** [review]

File: include/umf/providers/provider_os_memory.h [hunk 21]
```diff
@@ -81,6 +92,11 @@ typedef struct umf_os_memory_provider_params_t {
     /// part size for interleave mode - 0 means default (system specific)
     /// It might be rounded up because of HW constraints
     size_t part_size;
+
+    /// ordered list of the partitions for split mode
```
[the remark this answers is not in the document]
Comment (lplewa) [reply include/umf/providers/provider_os_memory.h:99]: done

File: include/umf/providers/provider_os_memory.h [hunk 22]
```diff
@@ -81,6 +92,11 @@ typedef struct umf_os_memory_provider_params_t {
...
+    umf_numa_split_partition_t *partitions;
+    /// len of partitions array
```
[the remark this answers is not in the document]
Comment (lplewa) [reply include/umf/providers/provider_os_memory.h:101]: done

File: src/provider/provider_os_memory.c [hunk 23]
```diff
@@ -509,15 +608,176 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+        if (bind == 0) {
+            membind->node++;
+            assert(membind->node < provider->partitions_len);
+            bind_border_page = 1;
+        }
+    }
+
+    // Update bind size and remainder based on whether the binding crossed a partition boundary
+    if (bind_border_page) {
+        // this means that next page bellong to multiple nodes.
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: done

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp:428]: i added one to explain a tuple, and fixed test body - it should be easier to understand.

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: if not enouth numa nodes then test is skip
if more or equal we set array len of numa to this value, to limit numanodes to required amount.

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: It might - but i find it useful if someone wants to extend test.
It is not easy to create output array, and it is not easy to find out if there is just entry missing

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 24]
```diff
@@ -425,6 +425,165 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+            ret.push_back({z, z * (z - 1) / 2 + z, in3, out3});
+        }
+        return ret;
+    }
+};
+
+INSTANTIATE_TEST_SUITE_P(checkModeSplit, testNumaSplit,
+                         ::testing::ValuesIn(testNumaSplit::getTestInput()));
+
+TEST_P(testNumaSplit, checkModeSplit) {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp:538]: done

**Review by bratpiorka — APPROVED** [review]

File: src/provider/provider_os_memory.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory.c:603]: sounds like a good idea, in the meantime we could do better, piece by piece in each PR

nonetheless, I'd be fine with anything, but not this: `/// membbind_t - some` 😃

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:428]: thanks!

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: now it looks better, thanks for defining named params instead of these `get<>` calls.

The question still stands - what `Wrong test in data` mean here...? did you mean `wrong test's params` or, more precise `out array size does not match pages count`...?

File: src/provider/provider_os_memory.c [hunk 25]
```diff
@@ -509,15 +599,179 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
...
+        return;
+    }
+
+    // Reset the bitmap for next binding
+    hwloc_bitmap_zero(membind->bitmap);
+
+    // Flag to check if binding crosses partition boundaries
+    int bind_border_page = 0;
+    if (membind->leftover_bind != 0) {
+        // if we have move than page leftover from previous bind
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory.c:647]: `than page` -> `than a page` (pls apply 3x in this file)

Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory.c:647]: `move` -> `more`?

File: src/provider/provider_os_memory.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory.c]: not done, 3 times to fix in this file

File: include/umf/providers/provider_os_memory.h (hunk 14 above)
Comment (lukaszstolarczuk) [reply include/umf/providers/provider_os_memory.h]: thx!

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 26]
```diff
@@ -425,6 +425,175 @@ TEST_F(testNuma, checkModeInterleaveCustomPartSize) {
...
+};
+
+INSTANTIATE_TEST_SUITE_P(checkModeSplit, testNumaSplit,
+                         ::testing::ValuesIn(testNumaSplit::getTestInput()));
+
+// positive test for numa mode split
+TEST_P(testNumaSplit, checkModeSplit) {
+    auto param = GetParam();
+    size_t page_size = sysconf(_SC_PAGE_SIZE);
+    auto required_numa_nodes = std::get<0>(param);
```
Comment (igchor) [inline test/provider_os_memory_multiple_numa_nodes.cpp:541]: nit: `auto [required_numa_nodes, pages, in, out] = param;`

File: src/provider/provider_os_memory.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c]: done

File: src/provider/provider_os_memory.c (hunk 25 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c:647]: memory binding

File: src/provider/provider_os_memory.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c:603]: done

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: Done

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 26 above)
Comment (lplewa) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: done

File: src/provider/provider_os_memory.c (hunk 25 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c]: done

File: src/provider/provider_os_memory.c [hunk 27]
```diff
@@ -600,15 +683,179 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
     return 0;
 }
 
-static int get_membind(os_memory_provider_t *provider, size_t size) {
+/// membbind_t - a memory binding itertator
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory.c:686]: `itertator` misspell

File: src/provider/provider_os_memory.c [hunk 28]
```diff
@@ -355,12 +432,20 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
     }
     LOG_INFO("established HWLOC NUMA policy: %u", provider->numa_policy);
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c:433]: done. Others parameters are not important so much.

File: src/provider/provider_os_memory.c [hunk 29]
```diff
@@ -327,6 +332,78 @@ create_fd_for_mmap(umf_os_memory_provider_params_t *in_params,
...
+            provider->partitions[i].weight = 1;
+            provider->partitions[i].target = provider->nodeset[i];
+        }
+        provider->partitions_weight_sum = provider->partitions_len;
+    } else {
+        provider->partitions_weight_sum = 0;
+        for (unsigned i = 0; i < in_params->partitions_len; i++) {
+            provider->partitions[i].weight = in_params->partitions[i].weight;
+            provider->partitions[i].target =
+                provider->nodeset[in_params->partitions[i].target];
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/provider/provider_os_memory.c:399]: done

File: src/provider/provider_os_memory.c (hunk 27 above)
Comment (lplewa) [reply src/provider/provider_os_memory.c:686]: done

File: src/provider/provider_os_memory.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory.c]: thx

**Review by lukaszstolarczuk — COMMENTED** [review]
👍
